### PR TITLE
feat(firestore): add WithCommitResponseTo TransactionOption

### PR DESCRIPTION
**Pull request by galenwarren**
Adds a new ```TransactionOption``` -- ```GetCommitTime``` -- that allows the caller to specify the address of a ```time.Time``` where the commit time of the transaction should be written, upon successful commit.

### CODE REVIEW

**Comment by galenwarren** [comment]
@telpirion @enocom  I was just checking on this. Would either of you be able to provide a code review? Thanks.

**Comment by enocom** [comment]
Sorry for the slow response @galenwarren. Ownership of this library has changed in the past few months. Let me reach out to some folks and see who should be looking at this.

**Comment by galenwarren** [comment]
Thanks!

**Comment by enocom** [comment]
cc @bhshkh and @kolea2 who might know who should look at this.

**Comment by bhshkh** [comment]
Reviewing

**Comment by galenwarren** [comment]
@bhshkh Thanks, please let me know if you have any questions or suggestions.

**Comment by galenwarren** [comment]
@bhshkh Hi, just checking in here. Is there any update? Thanks.

**Comment by bhshkh** [comment]
Under discussion internally. Will update

**Comment by galenwarren** [comment]
@bhshkh Is there anything I can do here to help?

**Comment by galenwarren** [comment]
Is this one not going to get merged?

**Comment by codyoss** [comment]
@bhshkh is this something we want to support?

**Comment by galenwarren** [comment]
If anyone would like me to rebase, just let me know. Thanks!

**Comment by bhshkh** [comment]
Sorry for the very delayed response. Reviewing this

**Comment by bhshkh** [comment]
One problem I see with the current design is that it is not extendable. This is what the commit response looks like
```
// The response for [Firestore.Commit][google.firestore.v1.Firestore.Commit].
type CommitResponse struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	// The result of applying the writes.
	//
	// This i-th write result corresponds to the i-th write in the
	// request.
	WriteResults []*WriteResult `protobuf:"bytes,1,rep,name=write_results,json=writeResults,proto3" json:"write_results,omitempty"`
	// The time at which the commit occurred. Any read with an equal or greater
	// `read_time` is guaranteed to see the effects of the commit.
	CommitTime *timestamppb.Timestamp `protobuf:"bytes,2,opt,name=commit_time,json=commitTime,proto3" json:"commit_time,omitempty"`
}
```


Apart from CommitTime field, there is another field called WriteResults. If/when in the future, we want to support retrieving WriteResults, we will have to introduce another transaction option and so on and so forth for any more fields added to the above struct.

Another thing is that the name `GetCommitTime` does not look appropriate. Open to any alternatives to that.

**Comment by bhshkh** [comment]
To minimize friction, consider setting `Allow edits from maintainers` on the PR, which will enable project committers and automation to update your PR.

**Comment by bhshkh** [comment]
Found similar request: https://github.com/googleapis/google-cloud-go/issues/1412

**Comment by bhshkh** [comment]
> @bhshkh is this something we want to support?

Yes

**Comment by galenwarren** [comment]
@bhshkh Thanks for looking at this.

Regarding the design -- agreed, it would be better if RunTransaction returned both a result (which could hold the CommitTime and the WriteResults) and an error, instead of just an error. But that would be a breaking change. I suppose a new RunTransaction (RunTransaction2, RunTransactionEx?) method could be added with a new signature, would that be better?

I'm happy to enable ```Allow edits from maintainers``` but I can't seem to find the option to select. Should I see that in the right-hand column somewhere?

```WithCommitResponseTo``` sounds good to me.

**Comment by bhshkh** [comment]
> @bhshkh Thanks for looking at this.
> 
> Regarding the design -- agreed, it would be better if RunTransaction returned both a result (which could hold the CommitTime and the WriteResults) and an error, instead of just an error. But that would be a breaking change. I suppose a new RunTransaction (RunTransaction2, RunTransactionEx?) method could be added with a new signature, would that be better?
> 
> I'm happy to enable `Allow edits from maintainers` but I can't seem to find the option to select. Should I see that in the right-hand column somewhere?
> 
> `WithCommitResponseTo` sounds good to me.

Instead of having a new function, we could create a new transaction option as you originally suggested.
In the original design, *time.time was getting passed as argument to trasaction option. Instead, use *pb.CommitResponse as argument.

**Comment by galenwarren** [comment]
Regarding "Allow edits and access secrets my maintainers": Yes, that's what I tried before. I just don't get a checkbox to enable it anywhere. Could it be because this PR is old somehow? Not sure ... it's strange.

I see what you're suggesting about CommitResponse, makes sense. I'll make the changes.

**Comment by galenwarren** [comment]
@bhshkh Just pushed some changes, please let me know what you think. Thanks.

File: firestore/transaction.go
Comment (jba) [inline firestore/transaction.go:84]: The API is a bit confusing so provide a small example in the doc:

//    var cr firestore.CommitResponse
//    err := client.RunTransaction(...., firestore.WithCommitResponseTo(&cr))
//    if err != nil {...}
//    // CommitReponse can be accessed here.

Actually one thing I'm not clear about is, is it populated if the transaction fails? Can that be documented too?

**Review by jba — CHANGES_REQUESTED** [review]

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go:85]: That's fine, I can create two different structs, one for the option and one for the response. They'll have identical structure, fwiw.

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go:84]: It's only populated on success -- will document:

```
// on success, handle the commit response
if err == nil {
	for _, opt := range opts {
		opt.handleCommitResponse(commitResponse)
	}
}
```

**Comment by galenwarren** [comment]
@bhshkh I saw the checks re: missing comments, I'll add them in the next commit where I address the other feedback. Will either be today or tomorrow.

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go]: Done in the latest commit, please let me know if that looks ok.

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go]: Done.

**Comment by galenwarren** [comment]
@bhshkh I added the missing comments, so that should address the first failed check. I wasn't sure quite what to make of the second failed check, it wasn't clear to me what it was saying, and it seemed to be wrt bigquery. I did rebase again with this last commit so maybe I've pulled in some new code related to that ...

File: firestore/transaction.go
Comment (bhshkh) [reply firestore/transaction.go]: The example needs to be added to firestore/examples_test.go 
I think thats what @jba meant.

Rest of the changes LGTM.

Once @jba has approved, I'll approve it too

Comment (galenwarren) [reply firestore/transaction.go]: @bhshkh @bja Gotcha, my mistake. I can make that change. 

Though I notice that, now that I'm returning an unexported ```commitResponse``` from ```WithCommitResponseTo```, one of the checks is failing with:

> firestore/transaction.go:96:46: exported func WithCommitResponseTo returns unexported type firestore.commitResponse, which can be annoying to use

I can think of a couple solutions.
* Change ```commitResponse``` to ```CommitResponseOption``` (we're already using ```CommitResponse``` for the wrapper struct)
* Return ```TransactionOption``` from ```WithCommitResponseTo```

In the spirit of avoiding back-and-forths, do either of you have a preference here (or alternate suggestion)?

Comment (bhshkh) [reply firestore/transaction.go]: I would wait for @jba's review.

Comment (galenwarren) [reply firestore/transaction.go]: @jba I was just wondering if you prefer either option or have a better idea. I can make the changes today. Thanks.

Comment (galenwarren) [reply firestore/transaction.go]: @jba @bhshkh Just checking in on this one. If nobody has a better suggestion, I guess I'll go with the first option, changing the return type to ```CommitResponseOption```.

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go]: Changed to ```CommitResponseOptions``` in the latest commit.

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go]: @jba @bhshkh I updated the example, renamed the option, and rebased to main. Please let me know if that looks ok. Thanks.

**Comment by galenwarren** [comment]
@jba @bhshkh Just wondering if we're still moving forward here?

**Comment by bhshkh** [comment]
> @jba @bhshkh Just wondering if we're still moving forward here?
Thanks for the modifications!
Yes.
Waiting on @jba's review.

**Comment by galenwarren** [comment]
@bhshkh @jba Thanks, let me know if there's anything I need to do.

**Comment by jba** [comment]
Sorry I dropped the ball on this. Looking now.

File: firestore/transaction.go
Comment (jba) [inline firestore/transaction.go:79]: I don't understand why this can't be unexported, like `ro` and `maxAttempts`. Why does that give an error when the other two don't? If there's some sort of special case to a check, let's apply it here too.

File: firestore/transaction.go
Comment (jba) [inline firestore/transaction.go:68]: Document this without referring to protos. (Say what it actually is.)

**Review by jba — CHANGES_REQUESTED** [review]

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go:79]: I don't really know, I just know that this is how it fails:

> firestore/transaction.go:96:46: exported func WithCommitResponseTo returns unexported type firestore.commitResponse, which can be annoying to use

```ro``` isn't returned from an exported function, so that wouldn't seem to apply. ```maxAttempts``` is a typedef for ```int```, not a struct, maybe that's the difference?

Comment (galenwarren) [reply firestore/transaction.go:79]: @jba Looks like this error is coming is coming straight out of ```golint```. There are various discussions about this rule out there ([here](https://github.com/golang/lint/issues/465), [here](https://github.com/golang/lint/issues/210)). I don't see how it can be bypassed at the ```golint``` level.

This project (```google-cloud-go```) does seem to ignore some ```golint``` errors, i.e. [here:](https://github.com/googleapis/google-cloud-go/blob/d410907f3e52bcc64bd92e0a341777c1277a6418/.github/workflows/vet.sh#L47), but as far as I can tell adding a new exclusion there would affect the entire project. Would you like me to do that?

Comment (jba) [reply firestore/transaction.go:79]: Yes, add that line, except write `firestore.commitResponse`

Comment (galenwarren) [reply firestore/transaction.go]: Done.

File: firestore/transaction.go
Comment (galenwarren) [reply firestore/transaction.go]: Done.

**Review by jba — APPROVED** [review]

**Review by bhshkh — APPROVED** [review]

**Comment by galenwarren** [comment]
Thanks @jba and @bhshkh .

What's the next step here, I'm not quite sure what this means:

> Waiting on code owner review from googleapis/yoshi-go-admins

**Comment by jba** [comment]
@codyoss I think your team needs to do something here?

**Review by codyoss — APPROVED** [review]
approved for the vet change

**Comment by bhshkh** [comment]
> but I can't seem to find the option to selec

Side note: Found the reason. https://github.com/googleapis/google-cloud-go/pull/10824#issuecomment-2331991189

**Comment by galenwarren** [comment]
@bhshkh @jba @codyoss Thanks all.

The only things that still seem to be failing are API checks against other parts of the project, I'm guessing those are due to outdated code in this branch, so I kicked off a merge to main.

What are the next steps, anything I need to do?

**Comment by bhshkh** [comment]
Sorry for late response.
Forced kokoro run and enabled auto merge
Will monitor

**Comment by bhshkh** [comment]
Unable to merge. This branch is out-of-date with the base branch

**Comment by bhshkh** [comment]
> > I'm happy to enable `Allow edits from maintainers` but I can't seem to find the option to select.
> 
> Side note: Found the reason. [#10824 (comment)](https://github.com/googleapis/google-cloud-go/pull/10824#issuecomment-2331991189)

@galenwarren , can you please recreate this PR from personal fork so that I have permissions to update the branch ?

**Comment by galenwarren** [comment]
@bhshkh I'd really like to avoid recreating the PR if possible. I've invited you to (temporarily) be a collaborator (writer) on the project in my organization, does that get you what you need?

**Comment by bhshkh** [comment]
> @bhshkh I'd really like to avoid recreating the PR if possible. I've invited you to (temporarily) be a collaborator (writer) on the project in my organization, does that get you what you need?

Sorry, I can't accept the invite

**Comment by galenwarren** [comment]
@bhshkh Do you mean it doesn't work or you're not allowed, by policy or something like that?

**Comment by galenwarren** [comment]
@bhshkh Wasn't sure if you saw my suggestion that I just rebase to main for you?

**Comment by bhshkh** [comment]
I'm not allowed to. Please rebase. I'll actively monitor the PR. Thanks!

**Comment by galenwarren** [comment]
@bhshkh Understood, thanks. I just rebased to main.

**Comment by bhshkh** [comment]
Merged !

**Comment by galenwarren** [comment]
Excellent - thanks!

**Pull request merged**
